src/main/java/inf112/skeleton/app/model/GameModel.java

```diff
@@ -40,55 +53,103 @@ public Player getPlayer() {
         return this.player;
     }
 
-    public Deque<Location> doPhase(int phaseNumber, Deque<Location> phaseSteps) {
-        //Gdx.app.log(GameModel.class.getName(), Integer.toString(phaseNumber));
-        if (phaseNumber == 5) return phaseSteps;
-        Location loc = phaseSteps.getLast().copy();
-        IProgramCard card = player.getCardInProgrammingSlot(phaseNumber);
-        player.setCardinProgrammingSlot(phaseNumber, null); // empty the slot
-        if (card != null && !(card instanceof MoveForwardCard && tiledMapHandler.wallInPath(loc.copy()))) {
-            phaseSteps.add(card.instruction(loc.copy()));
+    public void endTurn() {
+        Location loc = robot.getLocation();
+
+        for (int i = 0; i < 5; i++) {
+            doCard(i, loc);
+            loc = updateLastLoc(loc, cardSteps.get(i));
+            doTiles(i, loc);
+            loc = updateLastLoc(loc, tileSteps.get(i));
         }
-        loc = phaseSteps.getLast().copy();
 
-        // Check if the location is outside the map
```

Comment:
Nothing happens when the robot goes outside the board. Is that intentional? I know the way I made it happen was kind of a hack so it might be for the better